views-react/src/main/java/io/micronaut/views/react/ReactViewsRenderer.java

```diff
@@ -0,0 +1,216 @@
+            contextPool.release(context);
+        }
+    }
+
+    /**
+     * Needs to be public to be callable from the JS side.
+     * @hidden
+     */
+    public class RenderCallback {
+        private final Set<URI> urlsToPrefetch = new HashSet<>();
+
+        private final HashMap<URI, byte[]> prefetchedData = new HashMap<>();
+
+        private final Writer responseWriter;
+        private final String componentName;
+        private final JSContext jsContext;
+
+        public RenderCallback(Writer responseWriter, String componentName, JSContext jsContext) {
+            this.responseWriter = responseWriter;
+            this.componentName = componentName;
+            this.jsContext = jsContext;
+        }
+
+        @HostAccess.Export
+        public Value recordPrefetch(String urlAsString) throws URISyntaxException {
+            URI url = new URI(urlAsString);
+
+            // If we already fetched it in a previous round, return the data now.
+            var alreadyFetched = getPrefetchedDataForURL(url);
+            if (alreadyFetched != null)
+                return alreadyFetched;
+
+            // Otherwise, queue it for fetching.
+            if (urlsToPrefetch.add(url))
+                LOG.trace("React SSR request for <{}/> attempted to fetch {}", componentName, url);
+
+            return null;
+        }
+
+        @org.jetbrains.annotations.Nullable
+        private Value getPrefetchedDataForURL(URI url) {
+            var result = prefetchedData.get(url);
+            if (result == null) return null;
+            var jsonParse = jsContext.polyglotContext.eval("js", "JSON.parse");
+            // TODO: Work out the story around result parsing in the absence of the fetcher?
+            return jsonParse.execute(new String(result, StandardCharsets.UTF_8));
+        }
+
+        @HostAccess.Export
+        public ProxyObject getPrefetchedData() {
+            // This is called at the start of each render cycle. We'll drain the queue and return the cache built up
+            // so far, which will be serialized to JSON and returned to the client. If the next render completes
+            // without any new URLs being queued, the result will be returned as-is, otherwise we'll be called again
+            // to do more fetching.
+            var workList = new ArrayList<>(urlsToPrefetch);
+            urlsToPrefetch.clear();
+            for (URI url : workList) {
+                UriBuilder builder = UriBuilder.of(url);
+                if (url.getHost() == null) {
+                    // A URL like '/foo/bar' is valid. TODO: dynamic port discovery?
+                    builder = builder.host("localhost").port(8080).scheme("http");
+                }
+                var req = HttpRequest.GET(builder.build())
+                    .header(CONTENT_ENCODING, "UTF-8")
+                    .header(ACCEPT, "application/vnd.github.v3+json, application/json")
+                    .header(USER_AGENT, "Micronaut React SSR Prerender");
```

Comment:
A little bit concerned by the security implications here, where is `urlsToPrefetch` defined? Are they external resources? Seems like we are pre-rendering JS from external sources, could that be a danger? What is the sandboxing model?

If the resource is local can it be retrieved without the the HTTP request the localhost?

Reply:
The JS doesn't come from external sources. It's all a part of the app, so is "trusted". 

Nonetheless I am working on enabling sandboxing, because of course JavaScript world has giant dependency trees where not everyone maybe is entirely trustworthy, so to protect against supply chain attacks some sandboxing is useful anyway.

> If the resource is local can it be retrieved without the HTTP request the localhost?

See my question in #help on Micronaut Discord. I'm sure it can be, and that would be much better, I just don't know how to do it. I'd need to turn a URL into the right bean and method, and then build a reflective invocation in the same way Micronaut does with a fake HTTP request and all the injected stuff. If you know how to do that, some example code would be great.

Reply:
where are these resources kept? If they are on the classpath or file system you can use the ResourceResolver API

Reply:
Well they are usually API endpoints. The idea here is that the template uses the same code to download data on the client as on the server. For example, consider a page that lets you browse the profiles of a list of users. The initial render should send everything including a pre-selected/pre-loaded first user, and then when the user clicks the list, the data of that next user should be loaded. In effect we are simulating what the browser would do, but server side, so optimizing out a round-trip. Thus we can't load the result from the disk. It has to be genuinely simulated, including details that aren't done yet like propagating authentication cookies or headers.

Basically I need a way to efficiently do an HTTP request "as if" it was from the same user that generated the existing one.

Reply:
hmm, not sure if we have an API for that other than what you are doing, the question is should be allowing URLs other than localhost? Seems dangerous to potentially make outgoing calls and what if the deployed environment doesn't allow them?

Reply:
Good question.

Frontend JS can and often does make calls to multiple microservices. You can reverse proxy them all of course, and the integrations are hopefully robust to calls failing so this isn't a critical need, but in principle yes you would need to contact non-localhost servers at least sometimes. It probably makes sense to have a configurable whitelist of URL patterns to allow.

Reply:
I think it would be possible to fetch through Micronaut, to allow policies to be applied. Certainly the fetch implementation we use can be customized by policy or redirected to other endpoints.

Food for thought, because JS apps can and do structure their fetching this way. Some serverless database products (i.e. Cloudflare D1) work via REST JSON APIs under the hood, so this is functionally the same for a React app as say, JDBC, if you squint

Reply:
so if we continue with this approach I would like to see some whitelist implementation. I think from a security perspective it is too risky to by default allow outgoing calls to servers defined by the client JS. This could lead to an avenue whereby some vulnerability on the JS side could penetrate the server.

Reply:
Of course, it'd be pretty nice to short-circuit the HTTP stack and just pass the returned bean directly into the JS context when that's possible. But I guess that can be an optimization punted to later. You'd need APIs to look up the method for a URL, figure out reflectively if it fits the constraints etc. And presumably do that at compile time.

Reply:
@mikehearn Wait, is this for `fetch` as in fetching external data? Or passing beans? I'll ping you to discuss